### PR TITLE
[SPARK-38311][SQL] Fix DynamicPartitionPruning/BucketedReadSuite/ExpressionInfoSuite under ANSI mode

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Fix the following test suites under ANSI mode:

- DynamicPartitionPruningV1SuiteAEOff
- DynamicPartitionPruningV1SuiteAEOn
- DynamicPartitionPruningV2SuiteAEOff
- DynamicPartitionPruningV2SuiteAEOn
- ExpressionInfoSuite
- BucketedReadSuite

To fix ExpressionInfoSuite, this PR changes the invalid examples under ANSI mode and add comments about the behaviors with ANSI on/off.
### Why are the changes needed?

To set up a new GA test job with ANSI mode on 

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

Manually turn on ANSI mode and test

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 1]
```diff
@@ -200,6 +200,7 @@ class ExpressionInfoSuite extends SparkFunSuite with SharedSparkSession {
       val clonedSpark = spark.cloneSession()
       // Coalescing partitions can change result order, so disable it.
       clonedSpark.sessionState.conf.setConf(SQLConf.COALESCE_PARTITIONS_ENABLED, false)
+      clonedSpark.sessionState.conf.setConf(SQLConf.ANSI_ENABLED, false)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:203]: do you know which functions have query examples that rely on ANSI off?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala]: Yeah, I just added a "ignore" list when ANSI is on

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 2]
```diff
@@ -2344,8 +2344,11 @@ case class DateDiff(endDate: Expression, startDate: Expression)
     copy(endDate = newLeft, startDate = newRight)
 }
 
+// scalastyle:off line.size.limit
 @ExpressionDescription(
-  usage = "_FUNC_(year, month, day) - Create date from year, month and day fields.",
+  usage = """_FUNC_(year, month, day) - Create date from year, month and day fields.
+    If the configuration `spark.sql.ansi.enabled` is false, the function returns NULL on invalid inputs. Otherwise, it will throw an error instead.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2350]: can you run `DESC FUNCTION` locally and see if the output is reasonable? IIRC we need to put a super long single-line string for the doc.

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: ok, updated

**Comment by gengliangwang** [comment]
Merging to master

**Pull request closed without merging**